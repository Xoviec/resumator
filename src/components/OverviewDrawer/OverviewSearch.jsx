import React, { useCallback, useMemo, useState } from "react";
import { Search } from "@material-ui/icons";
import { alpha, makeStyles } from "@material-ui/core/styles";
import { InputBase } from "@material-ui/core";
import { debounce } from "debounce";

const useStyles = makeStyles((theme) => ({
  input: {
    width: "100%",
    paddingLeft: 50,
    border: "1px solid",
  },
  search: {
    position: "relative",
    borderRadius: theme.shape.borderRadius,
    backgroundColor: alpha(theme.palette.common.black, 0.05),
    "&:hover": {
      backgroundColor: alpha(theme.palette.common.black, 0.1),
    },
    marginBottom: theme.spacing(2),
    width: "100%",
  },
  searchIcon: {
    padding: theme.spacing(0, 2),
    height: "100%",
    position: "absolute",
    pointerEvents: "none",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    left: 0,
  },
}));

export const OverviewSearch = ({ handleSearch }) => {
  const classes = useStyles();
  const [searchVal, setSearchVal] = useState("");

  const debouncedSearch = useMemo(() => {
    return debounce((searchValue) => {
      handleSearch(searchValue);
    }, 500);
  }, [handleSearch]);

  const doSearch = useCallback(
    (event) => {
      setSearchVal(event.target.value);
      debouncedSearch(event.target.value);
    },
    [debouncedSearch]
  );

  return (
    <div className={classes.search}>
      <div className={classes.searchIcon}>
        <Search />
      </div>
      <InputBase
        variant="outlined"
        className={classes.input}
        onChange={doSearch}
        value={searchVal}
        placeholder="Searchâ€¦"
      />
    </div>
  );
};
