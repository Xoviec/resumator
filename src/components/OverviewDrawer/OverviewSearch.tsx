import { useCallback, useMemo, useState, VoidFunctionComponent } from "react";
import { Search } from "@mui/icons-material";
import { alpha } from "@mui/material/styles";
import { InputBase } from "@mui/material";
import { debounce } from "debounce";
import { styled } from "@mui/system";

const StyledInput = styled(InputBase)(({ theme }) => ({
  width: "100%",
  paddingLeft: 50,
  border: "1px solid",
}));

const StyledSearch = styled("div")(({ theme }) => ({
  position: "relative",
  borderRadius: theme.shape.borderRadius,
  backgroundColor: alpha(theme.palette.common.black, 0.05),
  "&:hover": {
    backgroundColor: alpha(theme.palette.common.black, 0.1),
  },
  marginBottom: theme.spacing(2),
  width: "100%",
}));

const StyledSearchIcon = styled("div")(({ theme }) => ({
  padding: theme.spacing(0, 2),
  height: "100%",
  position: "absolute",
  pointerEvents: "none",
  display: "flex",
  alignItems: "center",
  justifyContent: "center",
  left: 0,
}));

interface OverviewSearchProps {
  onSearchChange: (searchTerms: string) => void;
}

export const OverviewSearch: VoidFunctionComponent<OverviewSearchProps> = ({
  onSearchChange,
}) => {
  const [searchVal, setSearchVal] = useState("");

  const debouncedSearch = useMemo(() => {
    return debounce((searchValue: string) => {
      onSearchChange(searchValue);
    }, 500);
  }, [onSearchChange]);

  const doSearch = useCallback(
    (event) => {
      setSearchVal(event.target.value);
      debouncedSearch(event.target.value);
    },
    [debouncedSearch]
  );

  return (
    <StyledSearch>
      <StyledSearchIcon>
        <Search />
      </StyledSearchIcon>
      <StyledInput onChange={doSearch} value={searchVal} placeholder="Searchâ€¦" />
    </StyledSearch>
  );
};
