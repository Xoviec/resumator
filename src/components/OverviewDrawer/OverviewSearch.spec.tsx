import { createTheme, ThemeProvider } from "@mui/material";
import { render, screen } from "@testing-library/react";
import { waitFor } from "@testing-library/dom";
import userEvent from "@testing-library/user-event";
import { FunctionComponent } from "react";
import { OverviewSearch } from "./OverviewSearch";

const ThemeProviderWrapper: FunctionComponent = ({ children }) => {
  const theme = createTheme({});
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

describe("OverviewSearch", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it("should render correctly", () => {
    render(
      <ThemeProviderWrapper>
        <OverviewSearch onSearchChange={jest.fn()} />
      </ThemeProviderWrapper>
    );

    expect(document.body).toMatchSnapshot();

    expect(screen.getByPlaceholderText(/Search/)).toBeInTheDocument();
  });

  it("should debounce and call onSearchChange only once", async () => {
    const onSearchChange = jest.fn();

    render(
      <ThemeProviderWrapper>
        <OverviewSearch onSearchChange={onSearchChange} />
      </ThemeProviderWrapper>
    );

    userEvent.type(screen.getByPlaceholderText(/Search/), "test-content");

    await waitFor(() => expect(onSearchChange).toHaveBeenCalledTimes(1));
    expect(onSearchChange).toHaveBeenCalledWith("test-content");
  });
});
