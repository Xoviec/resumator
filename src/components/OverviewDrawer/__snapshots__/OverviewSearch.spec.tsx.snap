// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverviewSearch should render correctly 1`] = `
<body>
  <div>
    <div
      class="css-agzw14"
    >
      <div
        class="css-3lmxau"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="SearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </div>
      <div
        class="MuiInputBase-root MuiInputBase-colorPrimary css-i9rz6w-MuiInputBase-root"
      >
        <input
          class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
          placeholder="Searchâ€¦"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</body>
`;
